src/main/java/org/opensearch/ad/InsightsJobProcessor.java

```diff
@@ -0,0 +1,621 @@
+            if (hits.length == 0 || hits.length < baseSource.size()) {
+                log
+                    .info(
+                        "Successfully parsed {} anomalies in time window {} to {}",
+                        allAnomalies.size(),
+                        executionStartTime,
+                        executionEndTime
+                    );
+
+                if (!allAnomalies.isEmpty()) {
+                    // Enrich detector metadata (names, indices) before correlation
+                    fetchDetectorMetadataAndProceed(allAnomalies, jobParameter, lockService, lock, executionStartTime, executionEndTime);
+                } else {
+                    log.info("No anomalies found in time window, skipping ML correlation");
+                    releaseLock(jobParameter, lockService, lock);
+                }
+
+                injectSecurity.close();
+                return;
+            }
+
+            // continue to next page
+            Object[] next = hits[hits.length - 1].getSortValues();
+            fetchPagedAnomalies(
+                baseSource,
+                next,
+                allAnomalies,
+                injectSecurity,
+                jobParameter,
+                lockService,
+                lock,
+                executionStartTime,
+                executionEndTime
+            );
+        }, e -> {
+            if (e.getMessage() != null && (e.getMessage().contains("no such index") || e.getMessage().contains("index_not_found"))) {
+                log.info("Anomaly results index does not exist yet (no anomalies recorded)");
+            } else {
+                log.error("Failed to query anomaly results", e);
+            }
+            injectSecurity.close();
+            releaseLock(jobParameter, lockService, lock);
+        }));
+    }
+
+    /**
+     * Process anomalies with ML Commons metrics correlation.
+     * 
+     * @param jobParameter The insights job
+     * @param lockService Lock service for releasing lock
+     * @param lock The acquired lock
+     * @param anomalies All collected anomalies
+     * @param detectorMetadataMap Detector metadata for insights generation
+     * @param executionStartTime Start of analysis window
+     * @param executionEndTime End of analysis window
+     */
+    private void processAnomaliesWithMLCommons(
+        Job jobParameter,
+        LockService lockService,
+        LockModel lock,
+        List<AnomalyResult> anomalies,
+        Map<String, DetectorMetadata> detectorMetadataMap,
+        Instant executionStartTime,
+        Instant executionEndTime
+    ) {
+        MLMetricsCorrelationInput input = MLMetricsCorrelationInputBuilder
+            .buildInput(anomalies, detectorMetadataMap, executionStartTime, executionEndTime);
+
+        log.info("Built correlation input: {} metrics × {} buckets", input.getNumMetrics(), input.getNumBuckets());
+
+        log.info("Matrix contents: {}", input.getMatrix());
```

Comment:
The matrix can be huge and may expose metric/entity values. Recommend DEBUG with truncation or only log dimensions.

Reply:
added it for testing purpose, will clean up